paddle/math/float16.h

```diff
@@ -0,0 +1,714 @@
+    return static_cast<uint16_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator int32_t() const {
+    return static_cast<int32_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator uint32_t() const {
+    return static_cast<uint32_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator int64_t() const {
+    return static_cast<int64_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator uint64_t() const {
+    return static_cast<uint64_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator float() const {
+    return fp16_impl::half_to_float(*this);
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator double() const {
+    return static_cast<double>(fp16_impl::half_to_float(*this));
+  }
+};
+
+// Arithmetic operators
+#if defined(PADDLE_CUDA_FP16) && defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 530
+__device__ inline float16 operator+(const float16& a, const float16& b) {
+  return float16(__hadd(half(a), half(b)));
+}
+
+__device__ inline float16 operator-(const float16& a, const float16& b) {
+  return float16(__hsub(half(a), half(b)));
+}
+
+__device__ inline float16 operator*(const float16& a, const float16& b) {
+  return float16(__hmul(half(a), half(b)));
+}
+
+__device__ inline float16 operator/(const float16& a, const float16& b) {
+  // TODO(kexinzhao): check the cuda version that starts to support __hdiv
+  float num = __half2float(half(a));
+  float denom = __half2float(half(b));
+  return float16(num / denom);
+}
+
+__device__ inline float16 operator-(const float16& a) {
+  return float16(__hneg(half(a)));
+}
+
+__device__ inline float16& operator+=(float16& a, const float16& b) {
+  a = a + b;
+  return a;
+}
+
+__device__ inline float16& operator-=(float16& a, const float16& b) {
+  a = a - b;
+  return a;
+}
+
+__device__ inline float16& operator*=(float16& a, const float16& b) {
+  a = a * b;
+  return a;
+}
+
+__device__ inline float16& operator/=(float16& a, const float16& b) {
+  a = a / b;
+  return a;
+}
+
+__device__ inline bool operator==(const float16& a, const float16& b) {
+  return __heq(half(a), half(b));
+}
+
+__device__ inline bool operator!=(const float16& a, const float16& b) {
+  return __hne(half(a), half(b));
+}
+
+__device__ inline bool operator<(const float16& a, const float16& b) {
+  return __hlt(half(a), half(b));
+}
+
+__device__ inline bool operator<=(const float16& a, const float16& b) {
+  return __hle(half(a), half(b));
+}
+
+__device__ inline bool operator>(const float16& a, const float16& b) {
+  return __hgt(half(a), half(b));
+}
+
+__device__ inline bool operator>=(const float16& a, const float16& b) {
+  return __hge(half(a), half(b));
+}
+
+// On ARMv8.2-A CPU
+#elif defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
+    (PADDLE_GNUC_VER >= 71 || PADDLE_CLANG_VER >= 39)
```

Comment:
> use of undeclared identifier 'vaddh_f16'

I did not found `arm_fp16.h` in the `android-ndk-r15c` which the clang compiler version is 5.0.
Did I miss something?

Reply:
Currently clang does not support armv8.2 float16 neon intrinsics. (The currently developing clang 6.0 is planning to add this support). So I add assembly code for float16 arithmetic operators on the armv8.2 architecture, which should work for both gcc and clang.